### PR TITLE
HBW-60 Replace jquery.formatter with custom formatter for numbers

### PR DESCRIPTION
**Pull request by sugarfree1**

### CODE REVIEW

**Comment by TimAle** [comment]
Maybe we can place it in separate module with interface like previous formatter had?

**Comment by sugarfree1** [comment]
Updated

File: hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee [hunk 1]
```diff
@@ -222,6 +247,7 @@ modulejs.define 'HBWFormString', ['React', 'jQuery', 'HBWCallbacksMixin', 'HBWDe
         fullValue = this.substitute(strippedParts, substitutions)
         patternedValue = substitutions.join('')
 
-        [patternedValue, fullValue]
+        result = CustomFormatter.applyMaskForValue(this.props.params.pattern, value, this.templateRegexp)
+        [patternedValue, fullValue, result[0], result[1]]
```
Comment (TimAle) [inline hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee]: Let's use objects not arrays for clarity sake.

File: hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee [hunk 2]
```diff
@@ -129,14 +132,36 @@ modulejs.define 'HBWFormString', ['React', 'jQuery', 'HBWCallbacksMixin', 'HBWDe
...
+
+      [@stripNonAlphanumericChars(value), CustomFormatter.applyMaskForValue(pattern, value, this.templateRegexp)]
+
+    updateValue: ($el, remove) ->
       if @extractValueRegexp
-        substitutions = this.getSubstitutions(@extractValueRegexp, $el.val())
-        value = this.substitute(@valueParts, substitutions)
-        newValue = @stripNonAlphanumericChars(value)
-        @setState(value: newValue)
+        visualValue = this.state.visualValue
```
Comment (TimAle) [inline hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee]: I see `this.` in some places and `@` in other. Do we have any convention about this sugar?

**Comment by sugarfree1** [comment]
Updated

File: hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee [hunk 3]
```diff
@@ -85,28 +77,39 @@ modulejs.define 'HBWFormString', ['React', 'jQuery', 'HBWCallbacksMixin', 'HBWDe
...
     onLoadValidation: ->
-      if @validationRequired() && !!this.state.value
+      if @validationRequired() && !!@state.value
         @controlValidationTooltip(@isValid())
         @setValidationState()
 
     onChange: (event) ->
-      $el = jQuery(event.target)
-      @updateValue($el)
+      @updateValue(jQuery(event.target)[0], false)
```
Comment (TimAle) [inline hbw/app/assets/javascripts/hbw/components/form/string.js.jsx.coffee]: It seems we don't need to pass jQuery object to `updateValue`. `jQuery(event.target)[0]` --> `event.target`

File: vendor/assets/javascripts/formatter.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+var CustomFormatter = {};
+
+CustomFormatter.getNextValForAdd = function (prevVal, nextVal, position, pattern) {
+  if (prevVal[position] == ' ' && nextVal.length <= pattern.length) {
+    nextVal = prevVal.substr(0, position) + nextVal[position] + prevVal.substr(position + 1, prevVal.length);
```
Comment (TimAle) [inline vendor/assets/javascripts/formatter.js]: Parameter reassignment.

File: vendor/assets/javascripts/formatter.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  return nextVal;
+};
+
+CustomFormatter.simplifyMask = function (mask, templateRegexp) {
+  var matches = mask.match(templateRegexp);
+
+  for (var i in matches) {
+    newPart = matches[i].replace('{{', '').replace('}}', '').replace(/9/g, '_');
```
Comment (TimAle) [inline vendor/assets/javascripts/formatter.js]: Undeclared variable.

File: vendor/assets/javascripts/formatter.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    mask = mask.replace(matches[i], newPart);
+  }
+
+  return mask;
+};
+
+CustomFormatter.applyMaskForValue = function (mask, value, templateRegexp) {
+  mask = CustomFormatter.simplifyMask(mask, templateRegexp);
```
Comment (TimAle) [inline vendor/assets/javascripts/formatter.js]: Parameter reassignment.

**Comment by sugarfree1** [comment]
Updated

File: vendor/assets/javascripts/formatter.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    if (i < 0)
+      i = simpleMask.length;
+
+    return {mask: simpleMask.replace(/_/g, ' '), position: i}
+  }
+};
+
+CustomFormatter.getNextValForRemove = function (pattern, templateRegexp, position, nextVal) {
+  var simpleMask = CustomFormatter.simplifyMask(pattern, templateRegexp);
+  var updateNextVal = nextVal;
```
Comment (TimAle) [inline vendor/assets/javascripts/formatter.js]: Typo?

**Comment by sugarfree1** [comment]
Updated

**Pull request merged**